validator_test.go

```diff
@@ -0,0 +1,197 @@
+				msg:              Message{Type: CreateMessageType},
+			},
+		},
+		// Failure case
+		{
+			description: "known message type with failing Validators",
+			value: Test{
+				m: map[MessageType]Validators{
+					SimpleEventMessageType: {AlwaysInvalidMsg},
+				},
+				msg: Message{Type: SimpleEventMessageType},
+			},
+			expectedErr: ErrInvalidMsgType,
+		},
+		{
+			description: "unknown message type without provided default Validator",
+			value: Test{
+				m: map[MessageType]Validators{
+					SimpleEventMessageType: {alwaysValidMsg},
+				},
+				msg: Message{Type: CreateMessageType},
+			},
+			expectedErr: ErrInvalidMsgType,
+		},
+	}
+	for _, tc := range tests {
+		t.Run(tc.description, func(t *testing.T) {
+			assert := assert.New(t)
+			msgv, err := NewMsgTypeValidator(tc.value.m, tc.value.defaultValidator)
+			assert.NotNil(msgv)
+			assert.Nil(err)
+			err = msgv.Validate(tc.value.msg)
+			if tc.expectedErr != nil {
+				assert.ErrorIs(err, tc.expectedErr)
+				return
+			}
+
+			assert.Nil(err)
+		})
+	}
+}
+
+func testNewMsgTypeValidator(t *testing.T) {
+	type Test struct {
+		m                map[MessageType]Validators
+		defaultValidator Validator
+	}
+
+	var alwaysValidMsg ValidatorFunc = func(msg Message) error { return nil }
+	tests := []struct {
+		description string
+		value       Test
+		expectedErr error
+	}{
+		// Success case
+		{
+			description: "with provided default Validator",
+			value: Test{
+				m: map[MessageType]Validators{
+					SimpleEventMessageType: {alwaysValidMsg},
+				},
+				defaultValidator: alwaysValidMsg,
+			},
+			expectedErr: nil,
+		},
+		{
+			description: "without provided default Validator",
+			value: Test{
+				m: map[MessageType]Validators{
+					SimpleEventMessageType: {alwaysValidMsg},
+				},
+			},
+			expectedErr: nil,
+		},
+		{
+			description: "empty list of message type Validators",
+			value: Test{
+				m: map[MessageType]Validators{
+					SimpleEventMessageType: {},
+				},
+				defaultValidator: alwaysValidMsg,
+			},
+			expectedErr: nil,
+		},
+		// Failure case
+		{
+			description: "missing message type Validators",
+			value:       Test{},
+			expectedErr: ErrInvalidMsgTypeValidator,
+		},
+	}
+	for _, tc := range tests {
+		t.Run(tc.description, func(t *testing.T) {
+			assert := assert.New(t)
+			msgv, err := NewMsgTypeValidator(tc.value.m, tc.value.defaultValidator)
+			assert.NotNil(msgv)
+			if tc.expectedErr != nil {
+				assert.ErrorIs(err, tc.expectedErr)
+				return
+			}
+
+			assert.Nil(err)
+		})
+	}
+}
+
+func testAlwaysInvalidMsg(t *testing.T) {
+	assert := assert.New(t)
+	msg := Message{}
+	err := AlwaysInvalidMsg(msg)
+
+	assert.NotNil(err)
+	assert.ErrorIs(err, ErrInvalidMsgType)
+
+}
+
+func TestHelperValidators(t *testing.T) {
+	tests := []struct {
+		name string
+		test func(*testing.T)
+	}{
+		{"AlwaysInvalidMsg", testAlwaysInvalidMsg},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, tc.test)
+	}
+}
+
+func TestMsgTypeValidator(t *testing.T) {
```

Comment:
@kristinapathak  Yeah, I did it so I can quickly see all the tests for that type and the stacked test output of the subtests was nice.
But I can totally see why it seems silly since the subtests has the type name as their prefix.

Do we usually not group tests like this?

Reply:
And you also get vscode ui test grouping as well
![image](https://user-images.githubusercontent.com/12867159/170499647-08846c4d-ae7f-4a5d-9947-5a9f5b548baa.png)